packages/pyright-internal/src/analyzer/program.ts

```diff
@@ -351,8 +387,50 @@ export class Program {
         return sourceFile;
     }
 
-    setFileOpened(filePath: string, version: number | null, contents: string, options?: OpenFileOptions) {
+    setFileOpened(
+        filePath: string,
+        version: number | null,
+        contents: TextDocumentContentChangeEvent[],
+        options?: OpenFileOptions
+    ) {
+        let chainedFilePath: string | undefined = options?.chainedFilePath;
```

Comment:
See comment above. This is way more complicated than necessary. Please consider moving CoW logic to SourceFile.